### ISSUE TITLE
'bca' bootstrap method broken by pandas 0.21.0

### ISSUE THREAD

**Issue by akavalar** [bashtage/arch#298]
See [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-with-list-with-missing-labels-is-deprecated).

This affects the `_loo_jackknife` function in two places: `args_copy.append(arg.iloc[items])` and `kwargs_copy[k] = v.iloc[items]`.

I fixed it by changing `iloc` to `loc`, but then pandas throws out a deprecation warning. Far from elegant, but gets the job done (for me).

**Comment by bashtage** [comment]
I think there is somethign wrong with the data you are passing.  These are iloc which only depend on integer position of the data frame or series.  Are you perhaps passing a very short (1 observation?) series?  Changing those to .loc is definitely not the correct fix (in general). 

Could you post the call you are using to initialize the bootstrap and then call BCa?

**Comment by akavalar** [comment]
Sorry for the late response, life got in the way...

Yeah, you're totally right - this has nothing to do with pandas. When I reported this issue I was busy juggling several things at a time and managed to mix up scikit-learn's and pandas' version numbers (skl 0.21.0 came out not that long ago), and since I recently switched platforms and reinstalled my python distro I was sure this was being caused by one of the packages that was updated given that my code worked just fine before. Also, the `.iloc/.loc` fix is actually not a fix as you correctly pointed out above, I just didn't check whether it works properly until after opening the issue - oddly enough, it works for cases where the `bca` method fails right now (see below), but craps out for all other cases where `bca` works just fine. Oh, and it's been a while since I've used pandas 0.21.0 :)

That being said, there *is* some sort of an issue with the `bca` method. Feel free to replicate it using my data & code below. In short:
- I can't run an A/A test
- I can't switch the variants (i.e. go from A/B to B/A test) if the lengths of series differ

As mentioned above, the code worked just fine on some datasets but then it started throwing out these weird errors, so I simply assumed it was caused by some sort of dependency mismatch. Now it looks like it's actually caused by the part that implements the `bca` method (all other methods work just fine for all 4 case in the code below, btw).

```
import pandas as pd
from arch.bootstrap import IndependentSamplesBootstrap

data = pd.read_csv('bootstrap_test_bca.csv')
control = data.control[data.control.isna()==False] # 59 data points
treatment = data.treatment[data.treatment.isna()==False] # 60 data points

repetitions = int(1e3)
confidence_level = 0.95

def mean_diff(x, y):
    return x.mean() - y.mean()

# A/B: doesn't work
bootstrap = IndependentSamplesBootstrap(treatment, control)
bootstrap.conf_int(mean_diff, reps=repetitions, method='bca', tail='two', size=confidence_level)

# B/A: works
bootstrap = IndependentSamplesBootstrap(control, treatment)
bootstrap.conf_int(mean_diff, reps=repetitions, method='bca', tail='two', size=confidence_level)

# A/A: doesn't work
bootstrap = IndependentSamplesBootstrap(control, control)
bootstrap.conf_int(mean_diff, reps=repetitions, method='bca', tail='two', size=confidence_level)

# B/B: doesn't work
bootstrap = IndependentSamplesBootstrap(treatment, treatment)
bootstrap.conf_int(mean_diff, reps=repetitions, method='bca', tail='two', size=confidence_level)
```

Get data from here (60-line CSV file with two columns): [bootstrap_test_bca.csv](https://www.dropbox.com/s/23lxin2umaf1l1c/bootstrap_test_bca.csv)

Thanks!

**Comment by bashtage** [comment]
A short answer: I don't think BCa is well defined for independent samples. The LOO jackknife drops a single row from all arrays and then reestimates the statistics.  Dropping a single row is not particularly meaningful for unequal sample sizes. 

This page https://courses.washington.edu/matlab1/Bootstrap_examples.html suggests that BCa isn't applicable when sample sizes aren't the same. 

I should raise for unequal sample sizes.

**Comment by akavalar** [comment]
Hm, I never thought about this, but yeah, makes sense. I'm going to assume this has been solved (or, better put, will be soon) and close the issue. Thanks for your help!

**Comment by bashtage** [comment]
Reopening to remind myself to add a check and an exception. Thanks for the detailed report.

**Comment by bashtage** [comment]
Closed by 520af662c0227bd3268f0e1ca2b1441f8a0e1552
